src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java

```diff
@@ -59,58 +56,25 @@ public void setBeanFactory(final BeanFactory beanFactory) throws BeansException
     }
 
     @Override
-    protected Object handleInvocation(Object proxy, Method method,
-            Object[] args) throws Throwable {
-        final String methodName = method.getName();
-        final String nameSpace = getNamespace(proxy);
-
-        final String key = ProxyUtils.dottify(nameSpace) + "."
-                + ProxyUtils.dottify(methodName);
+    protected Object handleInvocation(Object proxy, Method method, Object[] args) {
+        final String key = createKey(getClass(proxy), method);
         final Object result = getConfig(key);
         if (result == null) {
             LOG.warn("Configuration not found for key: {}", key);
             return null;
         }
-
-        final Class<?> declaredReturnType = method.getReturnType();
-
-        try {
-
-            Constructor<?> constructor;
-            if (declaredReturnType.isInstance(result)) {
-                return result;
-            } else if (declaredReturnType.isPrimitive()) {
-                final Class<?> resultClass = result.getClass();
-                constructor = Primitives.wrap(declaredReturnType)
-                        .getDeclaredConstructor(resultClass);
-            } else if (declaredReturnType.isEnum()) {
-                for (Object t : Arrays
-                        .asList(declaredReturnType.getEnumConstants())) {
-                    if (result.toString().equalsIgnoreCase(t.toString())) {
-                        return t;
-                    }
-                }
-                LOG.warn("Unable to map [{}] to enum type [{}].", result, declaredReturnType.getName());
-                return null;
-            } else {
-                constructor = declaredReturnType
-                        .getDeclaredConstructor(result.getClass());
-            }
-            return constructor.newInstance(result);
```

Comment:
Speaking about breaking changes. For complex configurations in my team we were using a constructor with `LinkedHashMap<String, ?> rawData` argument and then converted the payload to what we needed. Only LinkedHashMap worked.

Example:
```
{
     "alias": "feature.config.default.data",
     "defaultValue": {
      "a8a23682-1623-450b-8817-50c98827ea4e": [
        {
          "id": "default_data",
          "components": {
            "componentA": [
              {
                "targets": ["A", "B"],
                "data": {
                  "subdata1": "something 1",
                  "subdata2": "something 2",
                  "subdata3": [
                    {
                      "key1": "value1",
                      "key2": "value2"
                    },
                    {
                      "key1": "value3",
                      "key2": "value4"
                    }
                  ],
                  "media": {
                    "mobile": {
                      "src": "https://mobile-image.jpg"
                    },
                    "desktop": {
                     "src": "https://desktop-image.jpg"
                    }
                  },
                  ... some more fields here
                }
              }
...
```

Then the baigan interface looks like
```
@BaiganConfig
public interface FeatureConfig {
    FeatureConfigData defaultData();
}
```
And the actual class with data:
```
public class FeatureConfigData {

    private static final ObjectMapper MAPPER = JacksonConfig.getObjectMapper()
            .setPropertyNamingStrategy(PropertyNamingStrategies.LOWER_CAMEL_CASE);

    private final Map<UUID, List<FeatureConfig>> data;

    public FeatureConfigData(@NonNull LinkedHashMap<String, List<Map<String, ?>>> rawData) {
        data = MAPPER.convertValue(rawData, new TypeReference<>() {
        });
    }
```

The rest of the structure of inner classes is irrelevant as Jackson will take care of them. 


With this change, it looks like, these workarounds will fail, and hopefully, they will fail on build time, as team should have tests covering this configuration. But, I think, it is still important to announce the fact of the breaking change explicit somehow. 

I check an example on this PR, and looks like with some minor adjustments it's possible to still keep custom initialization logic, if needed, if  `FeatureConfigData` constructor will be annotated with `@JsonCreator`.

Reply:
Yes, I would think that everything should work fine if only `FeatureConfigData` can be properly deserialized by the object mapper. I also plan to make the object mapper configurable to make this more flexible.

You are right to point out that we have to highlight this when releasing, thanks a lot. We can work out what exactly to recommend when writing documentation and finalizing the change.

Reply:
`ObjectMapper' can now optionally be provided as bean `baiganObjectMapper`.

Reply:
> I check an example on this PR, and looks like with some minor adjustments it's possible to still keep custom initialization logic, if needed, if FeatureConfigData constructor will be annotated with @JsonCreator.

I made an end-2-end test and it looks like annotating the constructor with `@JsonCreator` is actually sufficient to make it work. I will not add this test to the PR as it is somewhat artificial. Generally, any class that the object mapper can deserialize from JSON can be used as type for a configuration.